lib/semantics/check-omp-structure.cc

```diff
@@ -148,9 +164,45 @@ void OmpStructureChecker::Enter(const parser::OmpLoopDirective::Simd &) {
   SetContextAllowedOnce(allowedOnce);
 }
 
+// 2.8.3 do-simd-clause -> do-clause |
+//                         simd-clause
+void OmpStructureChecker::Enter(const parser::OmpLoopDirective::DoSimd &) {
+  SetContextDirectiveEnum(OmpDirective::DO_SIMD);
+
+  OmpClauseSet allowed{OmpClause::PRIVATE, OmpClause::FIRSTPRIVATE,
+      OmpClause::LASTPRIVATE, OmpClause::LINEAR, OmpClause::REDUCTION,
+      OmpClause::ALIGNED};
+  SetContextAllowed(allowed);
+
+  OmpClauseSet allowedOnce{OmpClause::SCHEDULE, OmpClause::COLLAPSE,
+      OmpClause::ORDERED, OmpClause::SAFELEN, OmpClause::SIMDLEN};
+  SetContextAllowedOnce(allowedOnce);
+}
+
+// 2.11.4 parallel-do-simd-clause -> parallel-clause |
+//                                   do-simd-clause
+void OmpStructureChecker::Enter(
+    const parser::OmpLoopDirective::ParallelDoSimd &) {
+  SetContextDirectiveEnum(OmpDirective::PARALLEL_DO_SIMD);
+
+  OmpClauseSet allowed{OmpClause::DEFAULT, OmpClause::PRIVATE,
+      OmpClause::FIRSTPRIVATE, OmpClause::SHARED, OmpClause::COPYIN,
+      OmpClause::REDUCTION, OmpClause::LASTPRIVATE, OmpClause::LINEAR,
+      OmpClause::ALIGNED};
+  SetContextAllowed(allowed);
+
+  OmpClauseSet allowedOnce{OmpClause::IF, OmpClause::NUM_THREADS,
+      OmpClause::PROC_BIND, OmpClause::SCHEDULE, OmpClause::COLLAPSE,
+      OmpClause::ORDERED, OmpClause::SAFELEN, OmpClause::SIMDLEN};
+  SetContextAllowedOnce(allowedOnce);
+}
+
 void OmpStructureChecker::Leave(const parser::OmpClauseList &) {
   // 2.7 Loop Construct Restriction
-  if (GetContext().directive == OmpDirective::DO) {
+  if (GetContext().directive == OmpDirective::DO ||
+      GetContext().directive == OmpDirective::PARALLEL_DO ||
+      GetContext().directive == OmpDirective::DO_SIMD ||
+      GetContext().directive == OmpDirective::PARALLEL_DO_SIMD) {
```

Comment:
You need to do this a lot (i.e. check if the current directive is one of a set), so it might be worth writing a function for that.  Something like this:
```
bool IsContextDirective(const std::initializer_list<OmpDirective> &directives) {
  auto it{std::find(directives.begin(), directives.end(), GetContext().directive)};
  return it != directives.end();
}
```
Then use it like this:
```
if (IsContextDirective({OmpDirective::DO, OmpDirective::PARALLEL_DO,
        OmpDirective::DO_SIMD, OmpDirective::PARALLEL_DO_SIMD})) {
```

Maybe name it `IsContextDirectiveOneOf` if you want to be more explicit.

Reply:
There's also `lib/common/enum-set.h` which provides APIs like `set` and `bitset` to sets of enums.

Reply:
The `EnumSet` approach is especially good if you want to give the sets names so that you can use them in more than one place. For example, the set of DO, PARALLEL_DO, DO_SIMD, and PARALLEL_DO_SIMD is being used twice currectly. If there is a meaningful name you can give this set, you can define it once and use it in both places.

Also, there is already a name for this enum set type, so a set can be defined as:
`static constexpr OmpDirectiveSet thisSet{...};`

Reply:
Thank you all for the very detailed suggestions!

Reply:
Addressed in update commit.